src/Symfony/Component/HttpClient/Response/AmpResponse.php

```diff
@@ -0,0 +1,410 @@
+                $request->removeHeader('cookie');
+                $request->removeHeader('host');
+            }
+
+            $response = yield $multi->request($options, $request, $canceller->getToken(), $info, $onProgress, $handle);
+            $info['redirect_time'] = microtime(true) - $info['start_time'];
+        }
+    }
+
+    private static function addResponseHeaders(Response $response, array &$info, array &$headers): void
+    {
+        $info['http_code'] = $response->getStatus();
+
+        if ($headers) {
+            $info['debug'] .= "< \r\n";
+        }
+
+        $h = sprintf('HTTP/%s %s %s', $response->getProtocolVersion(), $response->getStatus(), $response->getReason());
+        $info['debug'] .= "< {$h}\r\n";
+        $info['response_headers'][] = $h;
+
+        $headers = method_exists($response, 'getRawHeaders') ? $response->getRawHeaders() : $response->getHeaders();
+
+        foreach ($headers as $name => $values) {
+            foreach ($values as $v) {
+                if (\is_array($v)) {
+                    [$name, $v] = $v;
+                }
+                $h = ucwords($name, '-').': '.$v;
+                $info['debug'] .= "< {$h}\r\n";
+                $info['response_headers'][] = $h;
+            }
+        }
+
+        $info['debug'] .= "< \r\n";
+    }
+
+    /**
+     * Accepts pushed responses only if their headers related to authentication match the request.
+     */
+    private static function getPushedResponse(Request $request, AmpClientState $multi, array &$info, array &$headers, array $options, ?LoggerInterface $logger)
+    {
+        if ('' !== $options['body']) {
+            return null;
+        }
+
+        $authority = $request->getUri()->getAuthority();
+
+        foreach ($multi->pushedResponses[$authority] ?? [] as $i => [$pushedUrl, $pushedRequest, $pushedResponse, $parentOptions]) {
+            if ($info['url'] !== $pushedUrl || $info['http_method'] !== $pushedRequest->getMethod()) {
+                continue;
+            }
+
+            foreach ($parentOptions as $k => $v) {
+                if ($options[$k] !== $v) {
+                    continue 2;
+                }
+            }
+
+            foreach (['authorization', 'cookie', 'range', 'proxy-authorization'] as $k) {
+                if ($pushedRequest->getHeaderArray($k) !== $request->getHeaderArray($k)) {
+                    continue 2;
+                }
+            }
+
+            $response = yield $pushedResponse;
+
+            foreach ($response->getHeaderArray('vary') as $vary) {
+                foreach (preg_split('/\s*+,\s*+/', $vary) as $v) {
+                    if ('*' === $v || ($pushedRequest->getHeaderArray($v) !== $request->getHeaderArray($v) && 'accept-encoding' !== strtolower($v))) {
+                        $logger && $logger->debug(sprintf('Skipping pushed response: "%s"', $info['url']));
+                        continue 3;
+                    }
+                }
```

Comment:
You could use our helper functions for parsing these: https://github.com/amphp/http/blob/d2c00e3c48dc0511d2cea0d6fe73ea084f673aa4/src/functions.php#L5-L77

Reply:
Thanks for the link. I'm going to keep this one as is as it's quite simple, no key=value pairs here.